### ISSUE TITLE
Bug: Key Cloak

### ISSUE THREAD

**Issue by MeghanStothers** [bcgov/digital-journeys#277]

**Comment by MeghanStothers** [comment]
During a recent demo, I got the following message asking me to confirm my account

**Comment by MeghanStothers** [comment]
Tested pm Monday May 24 in
prod: 
https://digital-journeys.apps.silver.devops.gov.bc.ca/form/6283dbaab878d3863da71ef4
BUG REPORT: DID NOT WORK as I got the same message in PROD

Worked in DEV

**Comment by iman-jamali-fw** [comment]
@MeghanStothers, Can you please see if this is the error you get on prod? Just making sure I can replicate it. <br>
![Screen Shot 2022-05-30 at 10.18.39 AM.png](https://images.zenhubusercontent.com/61ddc48547f476ed3f35ca28/0d22c87e-302f-4426-bfee-3a7efadbab1c
)

**Comment by MeghanStothers** [comment]
Prod Link: https://digital-journeys.apps.silver.devops.gov.bc.ca/

Test Link https://digital-journeys-test.apps.silver.devops.gov.bc.ca/form

Dev Link: https://digital-journeys-dev.apps.silver.devops.gov.bc.ca/form

**Comment by MeghanStothers** [comment]
TESTED just now on Prod - I got the same message "Account already exists"

**Comment by iman-jamali-fw** [comment]
Logging with another IDIR subsequently will prompt you with the error (image attached below) and doesn't let you in.
Steps to reproduce:
1. login with an IDIR (let's call it idir1) => should work fine
2. Login with another IDIR (call it idir2) => show error "invalidFederatedIdentityActionMessage" and stop you from logging

Workaround: The error persists unless page cookies and cache are clear. This can be done through "clear site data" on chrome dev tools under the application tab.

The same thing happens if logging in with idir2 first, idir1 shows an error unless "clear site data" is clicked on chrome.

Affected environments: All

![Screen Shot 2022-05-31 at 2.04.59 PM.png](https://images.zenhubusercontent.com/61ddc48547f476ed3f35ca28/333157d2-2164-4a2a-ad47-eb96387a2d4f)

**Comment by MeghanStothers** [comment]
Will not hold up testing as we'll be testing in Test Environment 
Iman to write script with my IDIR 
Iman to ask me to test when he's ready

**Comment by MeghanStothers** [comment]
Tested the prod link and did not get the error. Iman to document the fix and then we can close this ticket - yay!

**Comment by iman-jamali-fw** [comment]
**What was the issue?**
In short, the issue was the broken identity provider link for a user. This prevented that user from login to the system.
![Screen Shot 2022-06-03 at 6.59.34 AM.png](https://images.zenhubusercontent.com/61ddc48547f476ed3f35ca28/b1d28668-cc93-47c7-baa0-25c11c355cd7)

**How to replicate it?** 
Logging to Keycloak and select one of the users. Go to `identity provider link` tab and remove the `ocio` link. The next time this user logs in will have the reported issue.

**More details on how the issue was fixed.**

Keycloak is integrated with ocio (IDIR) for login in this project. Sign on with IDIR is controlled by a login flow called `First login flow`. 

Quote from [Keycloak docs](https://www.keycloak.org/docs/latest/server_admin/index.html#default-first-login-flow-authenticators)

> First login flow
> When users log in through identity brokering, Keycloak imports and links aspects of the user within the realm’s local database. When Keycloak successfully authenticates users through an external identity provider, two situations can exist:
> 
> 
The default login flow handles these two cases:

> Keycloak has already imported and linked a user account with the authenticated identity provider account. In this case, Keycloak authenticates as the existing user and redirects back to the application.
> 
> No account exists for this user in Keycloak. Usually, you register and import a new account into the Keycloak database, but there may be an existing Keycloak account with the same email address. Automatically linking the existing local account to the external identity provider is a potential security hole. You cannot always trust the information you get from the external identity provider.
> 

The default login flow was not able to handle the case when an existing Keycloak account had a broken identity provider link. This is what happened to Meghan's account. To fix that, I had to add an "execution" to the default login flow to automatically fix the broken account. 

This was done by copying the default `First Broker Login`. This was necessary to be able to edit the flow. The new login flow then assign to IDIR identity provider. With that, the new login flow was edited by adding a new branch to that flow called "Automatically Set Existing User" to handle the broken existing account and automatically fix the broken link. More about "Automatically Set Existing User" from [Keycloak docs](https://www.keycloak.org/docs/latest/server_admin/index.html#_detect_existing_user_first_loging_flow)
![Screen Shot 2022-06-03 at 7.24.49 AM.png](https://images.zenhubusercontent.com/61ddc48547f476ed3f35ca28/82c4c29c-fd4e-435b-b735-6c2d8d76742b)
